httputil/httputil.go

```diff
@@ -26,33 +26,78 @@ type HttpClient interface {
 	Delete(url string, headers map[string]string, payload []byte) (responseCode int, body []byte, err error)
 }
 
+// for testing
+type httpClientInterface interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
 type Client struct {
-	httpClient    *http.Client
+	httpClient    httpClientInterface
 	retryBehavior RetryBehavior
 }
 
-type RetryBehavior = func(int) bool // return false to end retries
+type RetryBehavior = func(statusCode int, i int) bool
 
-var NoRetry = func(i int) bool {
+// return false to end retries
+// i starts from 1 keeps getting incremented while function returns true
+
+var NoRetry RetryBehavior = func(statusCode int, i int) bool {
 	return false
 }
 
-var LinearRetryThrice = func(i int) bool {
+var LinearRetryThrice RetryBehavior = func(statusCode int, i int) bool {
+	if !isTransientHttpStatusCode(statusCode) {
+		return false
+	}
 	time.Sleep(time.Second * 3)
 	if i < 3 {
 		return true // retry if count < 3
 	}
 	return false
 }
 
-var ExponentialRetryThrice = func(i int) bool {
-	time.Sleep(time.Second * time.Duration(3^(i+1)))
-	if i < 3 {
+var DefaultRetryBehavior = func(statusCode int, i int) bool {
```

Comment:
you might want to consider adding a comment that this is exponential retry behavior. It was obvious at first glance before, but isn't anymore after renaming it.

Reply:
Also, it seems you missed an opportunity here to declare DefaultRetryBehavior as a RetryBehavior type